File: app/src/main/java/tringapps/com/gearvrdemoifg/VrActivity.java
package tringapps.com.gearvrdemoifg;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.media.projection.MediaProjectionManager;
import android.os.Bundle;

import org.gearvrf.GVRActivity;
import org.gearvrf.GVRAndroidResource;
import org.gearvrf.GVRContext;
import org.gearvrf.GVRMain;
import org.gearvrf.GVRScene;
import org.gearvrf.GVRSceneObject;
import org.gearvrf.GVRTexture;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.json.JSONException;
import org.json.JSONObject;


public class VrActivity extends GVRActivity implements EventEndListener  {

    private static final int CAPTURE_PERMISSION_REQUEST_CODE = 1;
    private SampleMain main;
    @Override
    protected void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
    }

    @Override
    protected void onStop() {
        super.onStop();
        EventBus.getDefault().unregister(this);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(SocketMessage message) {
        if (message.getState().equals(SocketIO.GAME)){
            JSONObject game= (JSONObject) message.getMessage();
            try {
                if(game.getString(SocketIO.MESSAGE).equals(Constants.START)){
                    loadGame();
                }else if(game.getString(SocketIO.MESSAGE).equals(Constants.STOP)){
                    loadSplash();
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        main = new SampleMain(this, this);
        setMain(new Main());


        if(!Singleton.getInstance().isScreenCaptureStarted()) {
            Singleton.getInstance().setScreenCaptureStarted(true);
            startScreenCapture();
        }
    }

    @TargetApi(21)
    private void startScreenCapture() {

        MediaProjectionManager mediaProjectionManager = (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
        startActivityForResult(mediaProjectionManager.createScreenCaptureIntent(), CAPTURE_PERMISSION_REQUEST_CODE);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode,Intent data) {
        if (requestCode != CAPTURE_PERMISSION_REQUEST_CODE || resultCode != Activity.RESULT_OK)
            return;

        Intent cbIntent =  new Intent(this, MirroringService.class);
        cbIntent.putExtra(Constants.SCREEN_DATA, data);
        startService(cbIntent);
    }


    public void loadSplash(){
        getGVRContext().getMainScene().clear();
        main.loadSplash(getGVRContext());
    }

    public void loadGame(){
        getGVRContext().getMainScene().clear();
        main.loadGame(getGVRContext());
    }


    @Override
    public void changeMainScene(GVRScene mainScene) {
        getGVRContext().getMainScene().clear();
        main.loadBallon(getGVRContext());
    }

    private final class Main extends GVRMain {

        @Override
        public void onInit(GVRContext gvrContext) throws Throwable {

            //Load texture
            GVRTexture texture = gvrContext.getAssetLoader().loadTexture(new GVRAndroidResource(gvrContext, R.drawable.__default_splash_screen__));

            //Create a rectangle with the texture we just loaded
            GVRSceneObject quad = new GVRSceneObject(gvrContext, 4, 2, texture);
            quad.getTransform().setPosition(0, 0, -3);

            //Add rectangle to the scene
            gvrContext.getMainScene().addSceneObject(quad);


        }

        @Override
        public SplashMode getSplashMode() {
            return SplashMode.NONE;
        }

        @Override
        public void onStep() {
            //Add update logic here
        }
    }
}

File: app/src/main/java/tringapps/com/gearvrdemoifg/VrActivity.kt
package tringapps.com.gearvrdemoifg

import android.annotation.TargetApi
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjectionManager
import android.os.Bundle

import org.gearvrf.GVRActivity
import org.gearvrf.GVRScene
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.json.JSONException
import org.json.JSONObject

import tringapps.com.gearvrdemoifg.Constants.GAME
import tringapps.com.gearvrdemoifg.Constants.MESSAGE


class VrActivity : GVRActivity(), EventEndListener {

    companion object {
        private const val CAPTURE_PERMISSION_REQUEST_CODE = 1
    }

    private var main: SampleMain? = null
    private var screenCaptureIntent: Intent? = null
    private var screenCaptureResultCode: Int? = null

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(message: SocketMessage) {
        if (message.state == GAME) {
            val game = message.message as JSONObject?
            try {
                if (game != null)
                    when (game.getString(MESSAGE)) {
                        Constants.START -> loadGame()
                        Constants.STOP -> loadSplash()
                        Constants.INIT -> startMirroring()
                        Constants.TERMINATE -> stopMirroring()
                    }
            } catch (e: JSONException) {
                e.printStackTrace()
            }

        }
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        main = SampleMain(this, this)
        setMain(main)


        if (!MirroringService.isScreenCaptureStarted) {
            MirroringService.isScreenCaptureStarted = true
            startScreenCapture()
        }
    }

    override fun changeMainScene(mainScene: GVRScene) {
        gvrContext.mainScene.clear()
        main!!.loadBallon(gvrContext)
    }


    @TargetApi(21)
    private fun startScreenCapture() {
        val mediaProjectionManager = getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
        startActivityForResult(mediaProjectionManager.createScreenCaptureIntent(), CAPTURE_PERMISSION_REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (requestCode == CAPTURE_PERMISSION_REQUEST_CODE) {
            screenCaptureIntent = data
            screenCaptureResultCode = resultCode
        }

    }

    private fun startMirroring() {
        if (screenCaptureIntent != null && screenCaptureResultCode == Activity.RESULT_OK) {
            val cbIntent = Intent(this, MirroringService::class.java)
            cbIntent.putExtra(Constants.SCREEN_DATA, screenCaptureIntent)
            startService(cbIntent)
        }
    }

    private fun stopMirroring() {
        if (screenCaptureIntent != null && screenCaptureResultCode == Activity.RESULT_OK) {
            val cbIntent = Intent(this, MirroringService::class.java)
            stopService(cbIntent)
        }
    }

    private fun loadSplash() {
        gvrContext.mainScene.clear()
        main!!.loadSplash(gvrContext)
    }

    private fun loadGame() {
        gvrContext.mainScene.clear()
        main!!.loadBallon(gvrContext)
    }
}
